### PR TITLE
feat(composer): custom registries and basic auth

### LINKED ISSUE #2354: Composer support: private packages
**What would you like Renovate to be able to do?**

Since Composer support is underway (see #1357), it would be great if it would also support private package repositories, e.g. packages that aren't on Packagist or public GitHub repositories. Without that, Renovate Bot won't be able to update all PHP dependencies when Composer can't access these packages.

I first raised this at https://github.com/renovatebot/config-help/issues/73#issuecomment-410515712

**Describe the solution you'd like**

Private repositories are usually protected using HTTP authentication, for which the credentials can be stored in `~/.composer/auth.json` (see https://getcomposer.org/doc/articles/http-basic-authentication.md).

According to https://github.com/renovatebot/renovate/issues/1357#issuecomment-410938871 Renovate already creates a `auth.json` file for Composer.

What is now needed is a way to pass credentials for a private composer repository to Renovate that it then can add to `auth.json`.

### LINKED ISSUE #2574: Support custom register for composer
**What would you like Renovate to be able to do?**

Will be great if renovate will support custom registers.

Example:
```json
{
  "name": "wp-theme",
  "type": "wordpress-theme",
  "license": "MIT",
  "description": "Description.",
  "homepage": "https://insteria.com",
  "repositories": [
    {
      "type": "composer",
      "url": "https://wpackagist.org"
    }
  ],
  "extra": {
    "installer-paths": {
      "./plugins/{$name}/": [
        "type:wordpress-plugin"
      ],
    }
  },
  "require": {
    "php": "^7.1",
    "composer/installers": "~1.6.0",
    "wpackagist-plugin/breadcrumb-navxt": "~5.0.0"
  }
}
```

**Describe the solution you'd like**

I don't know.

**Describe alternatives you've considered**

I don't know.

**Additional context**

Doesn't exist.

### PR DESCRIPTION
**Pull request by rarkins**
Adds support for custom Packagist registries.

- [x] Support dynamic packages.json interface for wpackagist.org
- [x] Support static packages.json for satis
- [x] Support basic auth using hostRules
- [x] Write basic auth to `auth.json`

Closes #2574, Closes #2354

### CODE REVIEW

File: lib/datasource/packagist.js [hunk 1]
```diff
@@ -7,43 +7,105 @@ module.exports = {
...
+  );
+  return res;
+}
 
-  const pkgUrl = URL.resolve(regUrl, `/packages/${name}.json`);
+async function getAllPackages(regUrl) {
+  const registryMeta = await getRegistryMeta(regUrl);
+  const { providersUrl, files } = registryMeta;
+  const packages = {};
+  for (const file of files) {
```
Comment (rarkins) [inline lib/datasource/packagist.js]: TODO: run this in parallel

File: lib/datasource/packagist.js [hunk 2]
```diff
@@ -7,43 +7,105 @@ module.exports = {
...
+    cacheNamespace,
+    file.key,
+    { res, sha256 },
+    cacheMinutes
+  );
+  return res;
+}
 
-  const pkgUrl = URL.resolve(regUrl, `/packages/${name}.json`);
+async function getAllPackages(regUrl) {
```
Comment (rarkins) [inline lib/datasource/packagist.js]: Maybe cache this per-run

**Comment by rarkins** [comment]
This now appears to work with both packagist.org and wpackagist.org. Need a satis instance to test against

**Comment by rarkins** [comment]
TODO: honour `"packagist": false`

**Comment by Jamesking56** [comment]
Did you want me to set you up with a demo Satis repository?

**Comment by rarkins** [comment]
@Jamesking56 that would be perfect if you can.

**Comment by rarkins** [comment]
Now added (untested) code for basic auth

**Comment by Jamesking56** [comment]
> Did you want me to set you up with a demo Satis repository?

Made one, https://composer-demo.momentum.studio - Public gitlab repo for it: https://gitlab.com/momentumstudio/demo-composer

**Comment by rarkins** [comment]
Thanks, looks like I need to add some more code for that scenario but should be easy

**Comment by rarkins** [comment]
@Jamesking56 	Working now. Can you add basic auth to that URL - or a new one - so that I can test that too?

**Comment by Jamesking56** [comment]
Done, login:

User: `renovate`
Password: `renovate_password!`

**Comment by rarkins** [comment]
This is now functionally complete but without any unit tests yet. Tested against:

- packagist.org (no auth)
- wpackagist.org (no auth)
- Satis (basic auth)

**Comment by Jamesking56** [comment]
Great! Thanks for implementing this @rarkins :+1:

**Comment by Jamesking56** [comment]
I'm getting `error: unknown option `--hostRules'` on latest Docker container

**Comment by rarkins** [comment]
`—host-rules=`

**Comment by Jamesking56** [comment]
I am getting `INFO: Unauthorized Packagist repository` now when running:

````
docker run renovate/renovate:latest --host-rules='[{"endpoint":"https://composer.momentum.studio","username":"$SATIS_USERNAME","password":"$SATIS_PASSWORD"}]' --platform gitlab --timezone Europe/London --token $RENOVATE_TOKEN --autodiscover
````

**Comment by rarkins** [comment]
Are you sure the variable substitution is working? I thought single quotes like that don’t work

**Comment by Jamesking56** [comment]
Yeah got it working now I think thanks!

**Pull request merged**